File: src/song/timeline.ts
import type { Song, Pitch, Occurrence } from "./model.ts";
import { add, sub, mul, cmp, time, value, ZERO, type Time } from "./time.ts";
export interface BarSpan {
  id: string;
  sectionOccurrenceId: string;
  index: number;
  start: Time;
  length: Time;
  numerator: number;
  denominator: number;
  groups: number[];
}
export function bars(s: Song): BarSpan[] {
  const out: BarSpan[] = [];
  let start: Time = ZERO;
  for (const aid of s.arrangementOrder) {
    const a = s.tables.arrangement[aid]!;
    for (const id of s.tables.sections[a.sectionId]!.barIds) {
      const b = s.tables.bars[id]!;
      const length = b.actual ?? time(b.numerator * 4, b.denominator);
      out.push({
        id,
        sectionOccurrenceId: aid,
        index: out.length,
        start,
        length,
        numerator: b.numerator,
        denominator: b.denominator,
        groups: b.groups,
      });
      start = add(start, length);
    }
  }
  return out;
}
export const songEnd = (s: Song): Time => {
  let end: Time = ZERO;
  for (const b of bars(s)) end = add(b.start, b.length);
  for (const o of Object.values(s.tables.occurrences)) {
    const e = add(o.start, o.span);
    if (cmp(e, end) > 0) end = e;
  }
  return end;
};
interface Segment {
  start: Time;
  end: Time;
  phase: Time;
}
export function segments(s: Song, o: Occurrence): Segment[] {
  const end = add(o.start, o.span);
  const boundaries = bars(s)
    .filter(
      (b, i, a) =>
        i > 0 &&
        b.sectionOccurrenceId !== a[i - 1]!.sectionOccurrenceId &&
        cmp(b.start, o.start) > 0 &&
        cmp(b.start, end) < 0,
    )
    .map((b) => b.start);
  if (o.boundary === "continue" || boundaries.length === 0)
    return [{ start: o.start, end, phase: o.phase }];
  if (o.boundary === "stop")
    return [{ start: o.start, end: boundaries[0]!, phase: o.phase }];
  const points = [o.start, ...boundaries, end];
  return points
    .slice(0, -1)
    .map((p, i) => ({
      start: p,
      end: points[i + 1]!,
      phase: i === 0 ? o.phase : ZERO,
    }));
}
export interface Sound {
  id: string;
  occurrenceId: string;
  eventId: string;
  voiceId: string;
  partId: string;
  start: Time;
  duration: Time;
  pitch: Pitch | null;
  drum: "kick" | "snare" | "hat";
  gain: number;
  instrument: "guitar" | "bass" | "drums";
}
export function sounds(s: Song): Sound[] {
  const out: Sound[] = [];
  let iterations = 0;
  for (const o of Object.values(s.tables.occurrences)) {
    const pattern = s.tables.patterns[o.patternId]!;
    const voice = s.tables.voices[o.voiceId]!;
    const part = s.tables.parts[voice.partId]!;
    if (part.muted) continue;
    const events = Object.values(s.tables.events).filter(
      (e) => e.patternId === pattern.id,
    );
    for (const seg of segments(s, o))
      for (
        let cycle = sub(seg.start, seg.phase);
        cmp(cycle, seg.end) < 0;
        cycle = add(cycle, pattern.length)
      ) {
        if (++iterations > 100000)
          throw new Error(
            "Timeline is too dense to expand; shorten the selected occurrences",
          );
        for (const e of events) {
          if (e.kind === "rest") continue;
          const attack = add(cycle, e.start);
          const notes =
            e.kind === "chord"
              ? s.tables.chords[e.chordId!]!.notes.map((n) => ({
                  pitch: n.pitch,
                  member: n.id,
                }))
              : [{ pitch: e.kind === "drum" ? null : e.pitch, member: "" }];
          for (const n of notes) {
            const perf = e.performance.find((p) => p.memberId === n.member);
            const start = add(attack, perf?.offset ?? ZERO);
            if (cmp(start, seg.start) < 0 || cmp(start, seg.end) >= 0) continue;
            let duration = perf?.duration ?? e.duration;
            if (e.articulation === "staccato") duration = mul(duration, [1, 2]);
            if (o.tails === "cut" && cmp(add(start, duration), seg.end) > 0)
              duration = sub(seg.end, start);
            out.push({
              id: `${o.id}:${e.id}:${n.member}:${start}`,
              occurrenceId: o.id,
              eventId: e.id,
              voiceId: voice.id,
              partId: part.id,
              start,
              duration,
              pitch: n.pitch,
              drum: e.drum,
              instrument: part.instrument,
              gain:
                part.volume *
                e.accent *
                (e.articulation === "ghost" ? 0.25 : 1) *
                (e.articulation === "muted" ? 0.45 : 1),
            });
            if (out.length > 100000) throw new Error("Too many sounding notes");
          }
        }
      }
  }
  const rests = restSpans(s);
  const audible: Sound[] = [];
  for (const n of out) {
    let end = add(n.start, n.duration);
    let silent = false;
    for (const r of rests) {
      if (r.voiceId !== n.voiceId) continue;
      const restEnd = add(r.start, r.duration);
      if (cmp(n.start, r.start) >= 0 && cmp(n.start, restEnd) < 0) {
        silent = true;
        break;
      }
      if (cmp(r.start, n.start) > 0 && cmp(r.start, end) < 0) end = r.start;
    }
    if (!silent) audible.push({ ...n, duration: sub(end, n.start) });
  }
  return audible.sort((a, b) => cmp(a.start, b.start));
}
export function alignment(
  s: Song,
  ids: string[],
  after: Time,
  until: Time,
): Time | null {
  if (ids.length < 2) throw new Error("Select at least two occurrences");
  const sets = ids.map((id) => {
    const o = s.tables.occurrences[id];
    if (!o) throw new Error(`Unknown occurrence ${id}`);
    const times = new Map<string, Time>();
    let count = 0;
    for (const seg of segments(s, o))
      for (
        let t = sub(seg.start, seg.phase);
        cmp(t, seg.end) < 0 && cmp(t, until) <= 0;
        t = add(t, s.tables.patterns[o.patternId]!.length)
      ) {
        if (++count > 100000) throw new Error("Alignment range too dense");
        if (cmp(t, seg.start) >= 0 && cmp(t, after) > 0)
          times.set(t.join("/"), t);
      }
    return times;
  });
  for (const [key, t] of sets[0]!)
    if (sets.every((set) => set.has(key))) return t;
  return null;
}
export const secondsPerQuarter = (s: Song) =>
  60 / (s.tempo.bpm * value(s.tempo.beatUnit));
export function clicks(s: Song): { at: Time; strong: boolean }[] {
  return bars(s).flatMap((b) => {
    let offset: Time = ZERO;
    return b.groups.map((g, i) => {
      const at = add(b.start, offset);
      offset = add(offset, time(g * 4, b.denominator));
      return { at, strong: i === 0 };
    });
  });
}

export function cycleStarts(s: Song, o: Occurrence): Time[] {
  const starts: Time[] = [];
  let count = 0;
  for (const seg of segments(s, o))
    for (
      let at = sub(seg.start, seg.phase);
      cmp(at, seg.end) < 0;
      at = add(at, s.tables.patterns[o.patternId]!.length)
    ) {
      if (++count > 100000) throw new Error("Pattern is too dense");
      if (cmp(at, seg.start) >= 0) starts.push(at);
    }
  return starts;
}
export function restSpans(
  s: Song,
): { eventId: string; voiceId: string; start: Time; duration: Time }[] {
  const out: {
    eventId: string;
    voiceId: string;
    start: Time;
    duration: Time;
  }[] = [];
  let count = 0;
  for (const o of Object.values(s.tables.occurrences)) {
    const events = Object.values(s.tables.events).filter(
      (e) => e.patternId === o.patternId && e.kind === "rest",
    );
    if (!events.length) continue;
    for (const seg of segments(s, o))
      for (
        let at = sub(seg.start, seg.phase);
        cmp(at, seg.end) < 0;
        at = add(at, s.tables.patterns[o.patternId]!.length)
      ) {
        if (++count > 100000) throw new Error("Rest timeline is too dense");
        for (const e of events) {
          const start = add(at, e.start);
          if (cmp(start, seg.start) < 0 || cmp(start, seg.end) >= 0) continue;
          const end = add(start, e.duration);
          out.push({
            eventId: e.id,
            voiceId: o.voiceId,
            start,
            duration: cmp(end, seg.end) > 0 ? sub(seg.end, start) : e.duration,
          });
        }
      }
  }
  return out;
}

File: src-tauri/crates/song-core/src/timeline.rs
//! Timeline expansion: bars, segments, sounding notes, clicks and alignment.
//! Mirrors src/song/timeline.ts against the same validated song shapes.
//! Assumes a validated song; table lookups fail loudly instead of panicking.
use crate::{Placement, Result, Song, Time, ensure, lookup, placements};
use serde::Serialize;
use std::collections::HashSet;

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BarSpan {
    pub id: String,
    pub section_occurrence_id: String,
    pub index: usize,
    pub start: Time,
    pub length: Time,
    pub numerator: i32,
    pub denominator: i32,
    pub groups: Vec<i32>,
}

pub fn bars(song: &Song) -> Result<Vec<BarSpan>> {
    let mut out = Vec::new();
    let mut start = Time::ZERO;
    for aid in &song.arrangement_order {
        let arrangement = lookup(&song.tables.arrangement, "arrangement", aid)?;
        for id in &lookup(&song.tables.sections, "section", &arrangement.section_id)?.bar_ids {
            let bar = lookup(&song.tables.bars, "bar", id)?;
            let length = bar.actual.unwrap_or(Time::new(
                i64::from(bar.numerator) * 4,
                i64::from(bar.denominator),
            )?);
            out.push(BarSpan {
                id: id.clone(),
                section_occurrence_id: aid.clone(),
                index: out.len(),
                start,
                length,
                numerator: bar.numerator,
                denominator: bar.denominator,
                groups: bar.groups.clone(),
            });
            start = start.checked_add(length)?;
        }
    }
    Ok(out)
}

pub fn song_end(song: &Song) -> Result<Time> {
    let mut end = Time::ZERO;
    for bar in bars(song)? {
        let candidate = bar.start.checked_add(bar.length)?;
        if candidate > end {
            end = candidate;
        }
    }
    for placement in placements(song)? {
        let candidate = placement.start.checked_add(placement.span)?;
        if candidate > end {
            end = candidate;
        }
    }
    for region in song.tables.harmony.values() {
        if region.section_id.is_none() {
            let candidate = region.start.checked_add(region.duration)?;
            if candidate > end {
                end = candidate;
            }
        }
    }
    Ok(end)
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Segment {
    pub start: Time,
    pub end: Time,
    pub phase: Time,
}

pub fn segments(song: &Song, occurrence: &Placement) -> Result<Vec<Segment>> {
    let end = occurrence.start.checked_add(occurrence.span)?;
    let spans = bars(song)?;
    let mut boundaries = Vec::new();
    for (i, bar) in spans.iter().enumerate() {
        if i > 0
            && bar.section_occurrence_id != spans[i - 1].section_occurrence_id
            && bar.start > occurrence.start
            && bar.start < end
        {
            boundaries.push(bar.start);
        }
    }
    if occurrence.boundary == "continue" || boundaries.is_empty() {
        return Ok(vec![Segment {
            start: occurrence.start,
            end,
            phase: occurrence.phase,
        }]);
    }
    if occurrence.boundary == "stop" {
        return Ok(vec![Segment {
            start: occurrence.start,
            end: boundaries[0],
            phase: occurrence.phase,
        }]);
    }
    let mut points = vec![occurrence.start];
    points.extend(boundaries);
    points.push(end);
    let mut out = Vec::new();
    for (i, window) in points.windows(2).enumerate() {
        out.push(Segment {
            start: window[0],
            end: window[1],
            phase: if i == 0 { occurrence.phase } else { Time::ZERO },
        });
    }
    Ok(out)
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Sound {
    pub appearance_id: Option<String>,
    pub cycle_origin: Time,
    pub member_id: Option<String>,
    pub id: String,
    pub occurrence_id: String,
    pub event_id: String,
    pub voice_id: String,
    pub part_id: String,
    pub start: Time,
    pub duration: Time,
    pub pitch: Option<crate::Pitch>,
    pub drum: String,
    pub gain: f64,
    pub instrument: String,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RestSpan {
    pub event_id: String,
    pub voice_id: String,
    pub start: Time,
    pub duration: Time,
}

fn time_key(t: Time) -> String {
    let (n, d) = t.pair();
    format!("{n}/{d}")
}

pub fn sounds(song: &Song) -> Result<Vec<Sound>> {
    let mut out = Vec::new();
    let mut iterations = 0usize;
    for placement in placements(song)? {
        let pattern = lookup(&song.tables.patterns, "pattern", &placement.pattern_id)?;
        let voice = lookup(&song.tables.voices, "voice", &placement.voice_id)?;
        let part = lookup(&song.tables.parts, "part", &voice.part_id)?;
        if part.muted {
            continue;
        }
        let events: Vec<_> = song
            .tables
            .events
            .values()
            .filter(|e| e.pattern_id == pattern.id)
            .collect();
        for segment in segments(song, &placement)? {
            let mut cycle = segment.start.checked_sub(segment.phase)?;
            while cycle < segment.end {
                iterations += 1;
                ensure(
                    iterations <= 100000,
                    "Timeline is too dense to expand; shorten the selected occurrences",
                )?;
                for event in &events {
                    if event.kind == "rest" {
                        continue;
                    }
                    let attack = cycle.checked_add(event.start)?;
                    let mut notes = Vec::new();
                    if event.kind == "chord" {
                        let chord = lookup(
                            &song.tables.chords,
                            "chord",
                            event.chord_id.as_deref().unwrap_or(""),
                        )?;
                        for note in &chord.notes {
                            notes.push((Some(note.pitch.clone()), note.id.clone()));
                        }
                    } else {
                        let pitch = if event.kind == "drum" {
                            None
                        } else {
                            Some(event.pitch.clone())
                        };
                        notes.push((pitch, String::new()));
                    }
                    for (pitch, member) in notes {
                        let performance = event.performance.iter().find(|p| p.member_id == member);
                        let offset = performance.map_or(Time::ZERO, |p| p.offset);
                        let start = attack.checked_add(offset)?;
                        if start < segment.start || start >= segment.end {
                            continue;
                        }
                        let mut duration = performance.map_or(event.duration, |p| p.duration);
                        let articulation = match performance.and_then(|p| p.articulation.as_deref())
                        {
                            Some(a) if a != "inherit" => a,
                            _ => event.articulation.as_str(),
                        };
                        if articulation == "staccato" {
                            duration = duration.checked_mul(Time::new(1, 2)?)?;
                        }
                        if placement.tails == "cut" && start.checked_add(duration)? > segment.end {
                            duration = segment.end.checked_sub(start)?;
                        }
                        let (n, d) = start.pair();
                        out.push(Sound {
                            id: format!(
                                "{}:{}:{}:{member}:{n},{d}",
                                placement.id,
                                placement.appearance_id.as_deref().unwrap_or("global"),
                                event.id,
                            ),
                            member_id: if member.is_empty() {
                                None
                            } else {
                                Some(member)
                            },
                            occurrence_id: placement.id.clone(),
                            appearance_id: placement.appearance_id.clone(),
                            cycle_origin: start.checked_sub(event.start.checked_add(offset)?)?,
                            event_id: event.id.clone(),
                            voice_id: voice.id.clone(),
                            part_id: part.id.clone(),
                            start,
                            duration,
                            pitch,
                            drum: event.drum.clone(),
                            instrument: part.instrument.clone(),
                            gain: part.volume
                                * event.accent
                                * performance.and_then(|p| p.gain).unwrap_or(1.0)
                                * if articulation == "ghost" { 0.25 } else { 1.0 }
                                * if articulation == "muted" { 0.45 } else { 1.0 },
                        });
                        ensure(out.len() <= 100000, "Too many sounding notes")?;
                    }
                }
                cycle = cycle.checked_add(pattern.length)?;
            }
        }
    }
    let rests = rest_spans(song)?;
    let mut audible = Vec::new();
    for note in &out {
        let mut end = note.start.checked_add(note.duration)?;
        let mut silent = false;
        for rest in &rests {
            if rest.voice_id != note.voice_id {
                continue;
            }
            let rest_end = rest.start.checked_add(rest.duration)?;
            if note.start >= rest.start && note.start < rest_end {
                silent = true;
                break;
            }
            if rest.start > note.start && rest.start < end {
                end = rest.start;
            }
        }
        if !silent {
            let mut trimmed = note.clone();
            trimmed.duration = end.checked_sub(note.start)?;
            audible.push(trimmed);
        }
    }
    audible.sort_by_key(|note| note.start);
    Ok(audible)
}

pub fn alignment(song: &Song, ids: &[String], after: Time, until: Time) -> Result<Option<Time>> {
    ensure(ids.len() >= 2, "Select at least two occurrences")?;
    let all = placements(song)?;
    // Insertion order is ascending time; the first common time in the first
    // occurrence's order wins, so it must not be re-sorted by key spelling.
    let mut ordered: Vec<Vec<(String, Time)>> = Vec::new();
    let mut members: Vec<HashSet<String>> = Vec::new();
    for id in ids {
        let mut times = Vec::new();
        let mut seen = HashSet::new();
        let mut count = 0usize;
        for placement in all.iter().filter(|o| &o.id == id) {
            let pattern = lookup(&song.tables.patterns, "pattern", &placement.pattern_id)?;
            for segment in segments(song, placement)? {
                let mut at = segment.start.checked_sub(segment.phase)?;
                while at < segment.end && at <= until {
                    count += 1;
                    ensure(count <= 100000, "Alignment range too dense")?;
                    if at >= segment.start && at > after {
                        let key = time_key(at);
                        if seen.insert(key.clone()) {
                            times.push((key, at));
                        }
                    }
                    at = at.checked_add(pattern.length)?;
                }
            }
        }
        if times.is_empty() && !song.tables.occurrences.contains_key(id) {
            return Err(crate::Error::new(
                "invalid",
                format!("Unknown occurrence {id}"),
            ));
        }
        members.push(seen);
        ordered.push(times);
    }
    for (key, t) in &ordered[0] {
        if members.iter().all(|set| set.contains(key)) {
            return Ok(Some(*t));
        }
    }
    Ok(None)
}

pub fn seconds_per_quarter(song: &Song) -> f64 {
    60.0 / (song.tempo.bpm * song.tempo.beat_unit.value())
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Click {
    pub at: Time,
    pub strong: bool,
}

pub fn clicks(song: &Song) -> Result<Vec<Click>> {
    let mut out = Vec::new();
    for bar in bars(song)? {
        let mut offset = Time::ZERO;
        for (i, group) in bar.groups.iter().enumerate() {
            out.push(Click {
                at: bar.start.checked_add(offset)?,
                strong: i == 0,
            });
            offset = offset.checked_add(Time::new(
                i64::from(*group) * 4,
                i64::from(bar.denominator),
            )?)?;
        }
    }
    Ok(out)
}

pub fn cycle_starts(song: &Song, occurrence: &Placement) -> Result<Vec<Time>> {
    let mut starts = Vec::new();
    let mut count = 0usize;
    for segment in segments(song, occurrence)? {
        let pattern = lookup(&song.tables.patterns, "pattern", &occurrence.pattern_id)?;
        let mut at = segment.start.checked_sub(segment.phase)?;
        while at < segment.end {
            count += 1;
            ensure(count <= 100000, "Pattern is too dense")?;
            if at >= segment.start {
                starts.push(at);
            }
            at = at.checked_add(pattern.length)?;
        }
    }
    Ok(starts)
}

pub fn rest_spans(song: &Song) -> Result<Vec<RestSpan>> {
    let mut out = Vec::new();
    let mut count = 0usize;
    for placement in placements(song)? {
        let events: Vec<_> = song
            .tables
            .events
            .values()
            .filter(|e| e.pattern_id == placement.pattern_id && e.kind == "rest")
            .collect();
        if events.is_empty() {
            continue;
        }
        for segment in segments(song, &placement)? {
            let pattern = lookup(&song.tables.patterns, "pattern", &placement.pattern_id)?;
            let mut at = segment.start.checked_sub(segment.phase)?;
            while at < segment.end {
                count += 1;
                ensure(count <= 100000, "Rest timeline is too dense")?;
                for event in &events {
                    let start = at.checked_add(event.start)?;
                    if start < segment.start || start >= segment.end {
                        continue;
                    }
                    let end = start.checked_add(event.duration)?;
                    out.push(RestSpan {
                        event_id: event.id.clone(),
                        voice_id: placement.voice_id.clone(),
                        start,
                        duration: if end > segment.end {
                            segment.end.checked_sub(start)?
                        } else {
                            event.duration
                        },
                    });
                }
                at = at.checked_add(pattern.length)?;
            }
        }
    }
    Ok(out)
}
